go/vt/mysqlctl/xtrabackupengine.go

```diff
@@ -0,0 +1,484 @@
+	ctx context.Context,
+	cnf *Mycnf,
+	mysqld MysqlDaemon,
+	logger logutil.Logger,
+	dir string,
+	bhs []backupstorage.BackupHandle,
+	restoreConcurrency int,
+	hookExtraEnv map[string]string) (mysql.Position, error) {
+
+	var bh backupstorage.BackupHandle
+	var bm XtraBackupManifest
+	var toRestore int
+
+	for toRestore = len(bhs) - 1; toRestore >= 0; toRestore-- {
+		bh = bhs[toRestore]
+		rc, err := bh.ReadFile(ctx, backupManifest)
+		if err != nil {
+			log.Warningf("Possibly incomplete backup %v in directory %v on BackupStorage: can't read MANIFEST: %v)", bh.Name(), dir, err)
+			continue
+		}
+
+		err = json.NewDecoder(rc).Decode(&bm)
+		rc.Close()
+		if err != nil {
+			log.Warningf("Possibly incomplete backup %v in directory %v on BackupStorage (cannot JSON decode MANIFEST: %v)", bh.Name(), dir, err)
+			continue
+		}
+
+		logger.Infof("Restore: found backup %v %v to restore with %v file", bh.Directory(), bh.Name(), bm.FileName)
+		break
+	}
+	if toRestore < 0 {
+		// There is at least one attempted backup, but none could be read.
+		// This implies there is data we ought to have, so it's not safe to start
+		// up empty.
+		return mysql.Position{}, errors.New("backup(s) found but none could be read, unsafe to start up empty, restart to retry restore")
+	}
+
+	// Starting from here we won't be able to recover if we get stopped by a cancelled
+	// context. Thus we use the background context to get through to the finish.
+
+	logger.Infof("Restore: shutdown mysqld")
+	err := mysqld.Shutdown(context.Background(), cnf, true)
+	if err != nil {
+		return mysql.Position{}, err
+	}
+
+	logger.Infof("Restore: deleting existing files")
+	if err := removeExistingFiles(cnf); err != nil {
+		return mysql.Position{}, err
+	}
+
+	logger.Infof("Restore: reinit config file")
+	err = mysqld.ReinitConfig(context.Background(), cnf)
+	if err != nil {
+		return mysql.Position{}, err
+	}
+
+	// copy / extract files
+	logger.Infof("Restore: Extracting all files")
+
+	// first download the file into a tmp dir
+	// extract all the files
+	if err := be.restoreFile(ctx, cnf, logger, bh, !bm.SkipCompress, be.backupFileName()); err != nil {
+		logger.Errorf("error restoring backup file %v:%v", be.backupFileName(), err)
+		return mysql.Position{}, err
+	}
+
+	// copy / extract files
+	logger.Infof("Restore: Preparing the files")
+	// prepare the backup
+	restoreProgram := path.Join(*xtrabackupEnginePath, xtrabackup)
+	flagsToExec := []string{"--defaults-file=" + cnf.path,
+		"--prepare",
+		"--target-dir=" + cnf.TmpDir,
+	}
+	prepareCmd := exec.Command(restoreProgram, flagsToExec...)
+	prepareOut, _ := prepareCmd.StdoutPipe()
+	prepareErr, _ := prepareCmd.StderrPipe()
+	if err = prepareCmd.Start(); err != nil {
+		return mysql.Position{}, vterrors.Wrap(err, "unable to start prepare")
+	}
+
+	errOutput, _ := ioutil.ReadAll(prepareErr)
+	stdOutput, _ := ioutil.ReadAll(prepareOut)
+	err = prepareCmd.Wait()
+	if string(stdOutput) != "" {
+		// TODO remove after testing
```

Comment:
remove?

Reply:
will remove these comments, we decided to keep the logs.